File: DSAlgorithms/Leetcode/c++/2_pointers/189-rotate_arr.cpp
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        k = k % nums.size();      
        int left = 0; 
        int right = nums.size() - 1;
        while(left < right) //reverse entire arr
        {
            int temp = nums[left];
            nums[left] = nums[right];
            nums[right] =  temp;
            left ++;
            right --;
        }
        left = 0;
        right = k - 1;
        while(left < right) //reverse from start to k
        {
            int temp = nums[left];
            nums[left] = nums[right];
            nums[right] =  temp;
            left ++;
            right --;
        }
        left = k;
        right = nums.size() - 1;
        while(left < right) // reverse from k to end
        {
            int temp = nums[left];
            nums[left] = nums[right];
            nums[right] =  temp;
            left ++;
            right --;
        }
    }
};

File: DSAlgorithms/Leetcode/ts/src/2_pointers/189-rotate_arr.ts
/**
 Do not return anything, modify nums in-place instead.
 */
 const rotate = (nums: number[], k: number): void =>{
    k = k % nums.length;
    let left = 0;
    let right = nums.length - 1;
    while(left < right)
    {
        let temp = nums[left];
        nums[left] = nums[right];
        nums[right] = temp;
        left++;
        right--;
    }
    left = 0;
    right = k - 1;
    while(left < right)
    {
        let temp = nums[left];
        nums[left] = nums[right];
        nums[right] = temp;
        left++;
        right--;
    }
    left = k;
    right = nums.length - 1;
    while(left < right)
    {
        let temp = nums[left];
        nums[left] = nums[right];
        nums[right] = temp;
        left++;
        right--;
    }

};